README.md
rescuetime
==========

[![Maintainability](https://img.shields.io/codeclimate/maintainability/leesharma/rescuetime.svg?style=flat-square)](https://codeclimate.com/github/leesharma/rescuetime)
[![Coverage](https://img.shields.io/codeclimate/coverage/leesharma/rescuetime.svg?style=flat-square)](https://codeclimate.com/github/leesharma/rescuetime)
[![Build](https://img.shields.io/travis/leesharma/rescuetime.svg?style=flat-square)](https://travis-ci.org/leesharma/rescuetime)

[![Downloads](https://img.shields.io/gem/dt/rescuetime.svg?style=flat-square)](https://rubygems.org/gems/rescuetime)
[![Release](https://img.shields.io/github/release/leesharma/rescuetime.svg?style=flat-square)](https://github.com/leesharma/rescuetime/releases/)
[![License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](http://opensource.org/licenses/MIT)

A Ruby interface to the RescueTime APIs. Rescuetime provides a simple DSL for interacting
with your personal or team RescueTime data. Currently, this gem only supports the Data Analytics API with API key authorization.

For more information about RescueTime, visit [the RescueTime homepage](https://www.rescuetime.com).

#### README Navigation

* [Installation](#installation)
* [Usage](#usage)
   * [Prerequisites](#prerequisites)
   * [In a Nutshell](#in-a-nutshell) (skip to here if you want to see the syntax)
   * [Custom Report Formatters](#custom-report-formatters)
   * [Finding Answers (Documentation)](#finding-answers-documentation)
   * [Defaults](#defaults)
   * [Rescuetime Exceptions](#rescuetime-exceptions)
* [Development](https://github.com/leesharma/rescuetime/wiki/Development) ([section](#development))
* [Contributing](CONTRIBUTING.md) ([section](#contributing))
   * [TODO](#todo)

#### Useful Links
* [RDoc](http://www.rubydoc.info/gems/rescuetime)
* [Wiki](https://github.com/leesharma/rescuetime/wiki)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rescuetime'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rescuetime

## Usage

### Prerequisites

Ensure that you are using a [supported ruby version](https://github.com/leesharma/rescuetime/wiki/Supported-Rubies) for your project.

In order to use access your RescueTime data, you will need an API key. If you do not already have a key, visit the [API key management page](https://www.rescuetime.com/anapi/manage).

### In a Nutshell

```ruby
require 'rescuetime'

@client = Rescuetime::Client.new(api_key: <YOUR_API_KEY>)
@client.api_key?            #=> true
@client.valid_credentials?  #=> true

# Rescuetime uses lazy evaluation, so until you either manipulate the collection
# or explicitly call for it (with #all), it will remain in the Rescuetime::Collection
# format.
@client.overview.class      #=> Rescuetime::Collection
@client.overview.all.class  #=> Array
@client.overview.map {...}  #=> Array

@client.overview      # Returns an overview report, defaulting to "rank" order
@client.categories    # Returns a catigorical report, defaulting to "rank" order
@client.activities    # Returns a list of activities, defaulting to "rank" order
@client.productivity  # Returns a productivity report, defaulting to "rank" order
@client.efficiency    # Returns an efficiency report, defaulting to "time order"

##
# Date Range (:date, :frome, :to)
# -------------------------------
# Defaults:
#   If nothing is provided, defaults to current day (since 00:00)
#   If :from is provided, defaults :to to current day
#
# Valid date formats:
#   - "YYYY-MM-DD"    - "MM-DD-YYYY"    - "DD/MM"
#   - Object#strftime
@client.overview                      # Fetches results from today
@client.overview.date('2014-12-31')   # Fetches results from Dec 31, 2014.
@client.overview.from('2015-01-01').to('2015-02-01')
@client.overview.from('2015-04-01')


##
# Report Order (:order_by)
# ------------------------
# Defaults:
#   Efficiency defaults to chronological order; everything else defaults to "rank" order
#
# You can order_by:
#   :rank, :time, or :member (note: efficiency can't be sorted by :rank)
#
# When ordering by time, default interval is 1 hour.
# Options include:
#   :minute (5-minute chunks), :hour, :day, :week, :month
@client.efficiency    # Defaults to :time
@client.productivity  # Defaults to :rank

@client.productivity.order_by(:rank)
@client.productivity.order_by(:time)
@client.productivity.order_by(:member)

@client.productivity.order_by(:time)  # Defaults to :hour
@client.productivity.order_by(:time, interval: :minute)
@client.productivity.order_by(:time, interval: :hour)
@client.productivity.order_by(:time, interval: :day)
@client.productivity.order_by(:time, interval: :week)
@client.productivity.order_by(:time, interval: :month)

##
# Name Restrictions (:where)
# --------------------------
# Fetches results where name is an exact match
# The following reports can be limited by name:
#   :activities, :categories, :overview
#
# For activities, you can also limit by specific document title
#   (ex. document 'rails/rails' for activity 'github.com')
#   Try the query sans document for a list of valid options
#
# Names must be exact matches.
@client.activities.where(name: 'github.com')
@client.categories.where(name: 'Intelligence')
@client.overview.where(name: 'Utilities')
@client.activities.where(name: 'github.com', document: 'vcr/vcr')

##
# Formatting options (:csv, :array)
# ---------------------------------
@client.efficiency                  # Default return type is Array<Hash>
@client.efficiency.format(:cvs)     # Returns a CSV
@client.efficiency.format(:array)   # Returns Array<Hash>
```

### Custom Report Formatters

*Feature available in v0.4.0 and later*

Rescuetime ships with two report formats: CSV and Array. If you would like your
report in a different format, don't worry–it's easy to add a custom formatter.

Three things are required to add a custom formatter:

1. [Write a class](#writing-a-formatter) within the module
   `Rescuetime::Formatters` that inherits from `Rescuetime::Formatters::BaseFormatter`
   or one of its descendants
2. Define the class methods `.name` and `.format`
4. [Register your formatters](#registering-your-formatters) using `Rescuetime.configure`

#### Writing a Formatter

First, the formatters themselves. Here is a basic formatter:

```ruby
# config/formatters/nil_formatter.rb
module Rescuetime::Formatters
  # Turns a productivity report into nothing useful.
  class NilFormatter < BaseFormatter
    # @return [String]  name of this report format
    def self.name
      'nil'
    end

    # @param  [CSV] _report  the raw CSV report from Rescuetime
    # @return [nil]          the formatted output (in this case, nil)
    def self.format(_report)
      nil
    end
  end
end
```

You can even inherit from an existing formatter:

```ruby
# config/formatters/shouty_array_formatter.rb
module Rescuetime::Formatters
  # Formats a rescuetime report as an array of hashes, except shouting.
  class ShoutyArrayFormatter < ArrayFormatter
    # @return [String]  name of this report format
    def self.name
      'shouty_array'
    end

    # @param  [CSV]         report  the raw CSV report from Rescuetime
    # @return [Array<Hash>]         the formatted output (in this case, a shouty
    #                               array of hashes)
    def self.format(report)
      array = super(report)
      array.map do |hash|
        terms = hash.map { |key, value| [key.to_s.upcase, value.to_s.upcase] }
        Hash[terms]
      end
    end
  end
end
```

#### Registering your Formatters

Before setting your report format, add the path to your formatter(s) to the
Rescuetime configuration using the `Rescuetime.configure` method. You will be
able to set, append to, or manipulate the `formatter_paths` setting.

```ruby
Rescuetime.configure do |config|
  path = File.expand_path('../my_custom_formatter.rb', __FILE__)
  config.formatter_paths = [path]
end
```

Now Rescuetime will look for the `my_custom_formatter.rb` file. Multiple paths
may be added as well.

```ruby
Rescuetime.configure do |config|
  config.formatter_paths = [
    'config/formatters/*_formatter.rb',
    'lib/formatters/**/*_formatter.rb',
  ]
end
```

##### Rails Example

For example, in a Rails app, you could add the configuration file to `config/initializers`:

```ruby
# config/initializers/rescuetime.rb
Rescuetime.configure do |config|
  path = File.expand_path('../../formatters/*_formatter.rb', __FILE__)
  config.formatter_paths += [path]
end
```

Rails can now find any formatters ending in `_formatter.rb` in the folder `config/formatters`
(ex. `config/formatters/xml_formatter.rb`).

### Finding Answers (Documentation)

For more details, please see [official gem documentation](http://www.rubydoc.info/gems/rescuetime) or [read the wiki](https://github.com/leesharma/rescuetime/wiki).

### Defaults

The `Rescuetime::Client#activities` action has the following defaults:

```ruby

{ order_by:         'rank'
  interval:         'hour'
  date:             <TODAY> }

```

### Rescuetime Exceptions

*Note: In v0.4.0, custom exceptions were moved from the base `Rescuetime`
namespace to the `Rescuetime::Errors` namespace. For example, in v0.3.3,
a bad request was represented by `Rescuetime::BadRequest`. In >= v0.4.0,
it is represented by `Rescuetime::Errors::BadRequest`.*

There are a number of exceptions that extend from the `Rescuetime::Errors::Error` class:

* **`Rescuetime::Errors::MissingCredentialsError`** is raised when a request is attempted by a client with no credentials. Try setting credentials with `@client.api_key = <YOUR_API_KEY>`.
* **`Rescuetime::Errors::InvalidCredentialsError`** is raised when a request is attempted by a client with invalid credentials. Double-check your API key and fix your client with `@client.api_key = <VALID_API_KEY>`.
* **`Rescuetime::Errors::InvalidQueryError`** is raised if you enter an invalid value for any of the RescueTime query methods (or if the server returns a bad query error)
* **`Rescuetime::Errors::InvalidFormatError`** is raised if you pass a disallowed format to the client
* HTTP Response Errors:
    * **4xx => `Rescuetime::Errors::ClientError`**
    * 400 => `Rescuetime::Errors::BadRequest`
    * 401 => `Rescuetime::Errors::Unauthorized`
    * 403 => `Rescuetime::Errors::Forbidden`
    * 404 => `Rescuetime::Errors::NotFound`
    * 406 => `Rescuetime::Errors::NotAcceptable`
    * 422 => `Rescuetime::Errors::UnprocessableEntity`
    * 429 => `Rescuetime::Errors::TooManyRequests`
    * **5xx => `Rescuetime::Errors::ServerError`**
    * 500 => `Rescuetime::Errors::InternalServerError`
    * 501 => `Rescuetime::Errors::NotImplemented`
    * 502 => `Rescuetime::Errors::BadGateway`
    * 503 => `Rescuetime::Errors::ServiceUnavailable`
    * 504 => `Rescuetime::Errors::GatewayTimeout`

## Development

See the [development page](https://github.com/leesharma/rescuetime/wiki/Development) of the wiki.

## Contributing

See the [contributing page](CONTRIBUTING.md).

### TODO

The eventual goal of this gem is full coverage of all Rescuetime APIs.

Associated issues from our issue tracker are linked when applicable. If you are interested in contributing, please feel free to work on any of these issues! If there is a feature you want that's not on the TODO list, open an issue and we can discuss it.

- [x] Minimum viable functionality (in [v0.1.0])
- [x] Add support for major query params in Data Analytic API (in [v0.2.0])
- [x] Lazy query evaluation and method chaining (in [v0.3.0], [v0.3.1], [v0.3.2], [v0.3.3])
- [x] Add custom report formatters (in [v0.4.0])
- [ ] Add full coverage of Data Analytics API
  - [ ] Restrict results by group ([#14])
  - [ ] Restrict results by person
- [ ] Add support for "Daily Summary API" ([#28])
- [ ] Add support for "Alerts Feed API"
- [ ] Add support for "Daily Highlights Feed API"
- [ ] Add support for "Daily Highlights POST API"
- [ ] Add OAuth2 support

For details about completed items, please see [the release page][releases].

[#14]: https://github.com/leesharma/rescuetime/issues/14
[#28]: https://github.com/leesharma/rescuetime/issues/28

[v0.1.0]:https://github.com/leesharma/rescuetime/releases/tag/v0.1.0
[v0.2.0]:https://github.com/leesharma/rescuetime/releases/tag/v0.2.0
[v0.3.0]:https://github.com/leesharma/rescuetime/releases/tag/v0.3.0
[v0.3.1]:https://github.com/leesharma/rescuetime/releases/tag/v0.3.1
[v0.3.2]:https://github.com/leesharma/rescuetime/releases/tag/v0.3.2
[v0.3.3]:https://github.com/leesharma/rescuetime/releases/tag/v0.3.3
[v0.4.0]:https://github.com/leesharma/rescuetime/releases/tag/v0.4.0
[releases]:https://github.com/leesharma/rescuetime/releases


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

See also
- [GitHub releases](https://github.com/leesharma/rescuetime/releases)

## [Unreleased] - current

...

---

## [v1.0.1] - 2019-02-16

Mention max ruby version in gemspec

### Updates

No code updates. It looks like some functionality breaks under ruby 2.4+, so
this change updates the gemspec to reflect this. A future update will probably
resolve the bug.

## [v1.0.0] - 2016-08-26

Bump Version and Misc Updates

### Updates

The only update (aside from the version bump) is specifying that executables
are kept in the `bin` directory, not the `exe` directory. This version bump
is mainly to signify that the API is stable.

## [v0.4.0] - 2016-04-26

Add Custom Formatters and Misc Updates

### Breaking Changes

Exception classes, which were previously in the top-level Rescuetime (ex. Rescuetime::InvalidQueryError) are now under the Rescuetime::Errors module. This change was made for better code organization.

### Updates

#### Allow Frozen String Literals

If you're using ruby 2.3, rescuetime will now work with the `# frozen_string_literals: true` option set.

#### Custom Formatters

The big change in this update is that users can now add/configure custom formatters.

Rescuetime ships with two report formats: CSV and Array. If you would like your
report in a different format, don't worry–it's easy to add a custom formatter.

Three things are required to add a custom formatter:

    - Write a class within the module `Rescuetime::Formatters` that inherits from `Rescuetime::Formatters::BaseFormatter` or one of its descendants
    - Define the class methods `.name` and `.format`
    - Register your formatters using `Rescuetime.configure`

##### Writing a Formatter

First, the formatters themselves. Here is a basic formatter:

```ruby
# config/formatters/nil_formatter.rb
module Rescuetime::Formatters
  # Turns a productivity report into nothing useful.
  class NilFormatter < BaseFormatter
    # @return [String]  name of this report format
    def self.name
      'nil'
    end

    # @param  [CSV] _report  the raw CSV report from Rescuetime
    # @return [nil]          the formatted output (in this case, nil)
    def self.format(_report)
      nil
    end
  end
end
```

You can even inherit from an existing formatter:

```ruby
# config/formatters/shouty_array_formatter.rb
module Rescuetime::Formatters
  # Formats a rescuetime report as an array of hashes, except shouting.
  class ShoutyArrayFormatter < ArrayFormatter
    # @return [String]  name of this report format
    def self.name
      'shouty_array'
    end

    # @param  [CSV]         report  the raw CSV report from Rescuetime
    # @return [Array<Hash>]         the formatted output (in this case, a shouty
    #                               array of hashes)
    def self.format(report)
      array = super(report)
      array.map do |hash|
        terms = hash.map { |key, value| [key.to_s.upcase, value.to_s.upcase] }
        Hash[terms]
      end
    end
  end
end
```

##### Registering your Formatters

Before setting your report format, add the path to your formatter(s) to the
Rescuetime configuration using the Rescuetime.configure method. You will be
able to set, append to, or manipulate the formatter_paths setting.

```ruby
Rescuetime.configure do |config|
  path = File.expand_path('../my_custom_formatter.rb', __FILE__)
  config.formatter_paths = [path]
end
```

Now Rescuetime will look for the my_custom_formatter.rb file. Multiple paths
may be added as well.

```ruby
Rescuetime.configure do |config|
  config.formatter_paths = [
    'config/formatters/*_formatter.rb',
    'lib/formatters/**/*_formatter.rb',
  ]
end
```

### Documentation

Increased documentation coverage/quality

### Test Coverage

Increased

## [v0.3.3] - 2016-05-21

Remove the dependency to Faraday

### Refactoring

Switch from using Faraday to Net::HTTP in order to reduce production dependencies.

---

## [v0.3.2] - 2016-05-21

Documentation Comments and Refactoring

### Documentation

Public methods and classes now have 100% documentation coverage, and the documentation was expanded to include descriptions, examples, links, and more.

###Refactoring

The `Rescuetime::DateParser` class was extracted from the `Rescuetime::QueryBuilder`, keyword arguments were added in preference to the options hash, and other code readability/maintainability changes were made.

This release shouldn't affect the behavior or interface of the gem, and it adds no new features.

---

## [v0.3.1] - 2015-05-04

Fix gemspec description

### Minor fix:

gemspec description used `%w(...)`, changed to string

---

## [v0.3.0] - 2015-05-04

### Focus

Method Chaining and Lazy Evaluation

### Updates

This release introduces many breaking changes to the codebase. Since the gem is still in development (v0.x.x), the major version has not increased.

The major goal of this release was refactoring and adding method chaining and lazy evaluation. Additionally, results can now be limited by overview/category/activity name as well as document name,.

Queries can now be formed with the following syntax:

```ruby
client = Rescuetime::Client.new RESCUETIME_API_KEY
#=> <#Rescuetime::Client:...>

# Build the query
activities = client.activities.from(1.week.ago).where(name: 'github.com').format('csv')
#=> <#Rescuetime::Collection:...>

# Perform the query with #all or an enumerable method (#each, #map, #any?, etc.)
activities.all
#=> <#CSV:...>
```

---

## [v0.2.0] - 2015-04-23

### Scope

Provides nearly-full coverage of the RescueTime Analytic Data API.

### Features

```ruby
# Basics
@client = Rescuetime::Client.new

@client.api_key?            #=> false
@client.valid_credentials?  #=> false
@client.activities          #=> Rescuetime::MissingCredentials

@client.api_key='invalid'
@client.api_key?            #=> true
@client.valid_credentials?  #=> false

lib/rescuetime/activities.rb
module Rescuetime
  # The Rescuetime::Activities module contains client methods relating to user
  # activities on the /data endpoint of the Data Analytics API.
  #
  # @since v0.1.0
  module Activities
    # Base URL for RescueTime Data Analytics API endpoint
    # @since v0.1.0
    BASE_URL = 'https://www.rescuetime.com/anapi/data'
    # Map of numeric productivity levels to meaning
    # @since v0.2.0
    PRODUCTIVITY_LEVELS = { -2  => 'Very Unproductive',
                            -1  => 'Unproductive',
                            0   => 'Neutral',
                            1   => 'Productive',
                            2   => 'Very Productive' }

    # Returns array of all activities.
    #
    # @example Basic behavior
    #   @activities = @client.activities
    #   @activities.class     # => Array
    #   @acitvities[0].class  # => Hash
    #
    #   @activities[0]
    #   # =>  {
    #   #       :rank=>1,
    #   #       :time_spent_seconds=>5307,
    #   #       :number_of_people=>1,
    #   #       :activity=>"github.com",
    #   #       :category=>"General Software Development",
    #   #       :productivity=>2
    #   #     }
    #
    # @example Set level of detail
    #   @client.activities(detail: 'overview')[0]
    #   # => { :rank=>1, :time_spent_seconds=>13140, :number_of_people=>1, :category=>'Software Development' }
    #
    #   @client.activities(detail: 'category')[0]
    #   # => { :rank=>1, :time_spent_seconds=>5835, :number_of_people=>1, :category=>'Editing and IDEs' }
    #
    #   @client.activities(detail: 'activity')[0]
    #   # => { :rank=>1,
    #   #      :time_spent_seconds=>5835,
    #   #      :number_of_people=>1,
    #   #      :category=>'Editing and IDEs',
    #   #      :activity=>'RubyMine',
    #   #      :productivity=>2 }
    #
    #   @client.activities(detail: 'productivity')              # Equivalent to @client.productivity
    #   @client.activities(detail: 'efficiency', by: 'time')    # Equivalent to @client.efficiency
    #
    # @example Set perspective
    #   @client.activities(by: 'rank')     # Returns activities by rank (time spent per activity)
    #   @client.activities(by: 'interval') # Returns activities chronologically
    #   @client.activities(by: 'member')   # Returns activities grouped by member
    #
    # @example Set date range (note: no time constraints are allowed)
    #   @client.activities(date: '2014-05-06')                      # Returns report for May 6, 2014
    #   @client.activities(from: '2014-05-06', to: '2014-06-06')    # Returns report for May 6 to June 6, 2014
    #   @client.activities(from: '2015-04-10')                      # Returns report for April 10, 2015 to today
    #
    # @example Set time interval
    #   @client.efficiency( from: '2015-01-01',    # Returns an efficiency report by
    #                       to: Time.now,          # week from the Jan 1, 2015 through
    #                       interval: 'week')      # today.
    #
    # @example Format
    #   @client.activities                # Returns array of hashes
    #   @client.activities format: 'csv'  # Returns Mime::CSV
    #
    # @param [Hash] options Query parameters to be passed to RescueTime
    # @option options [String] :detail
    #   Restricts the level of detail of report returned
    #   1. 'overview': sums statistics for all activities into their top level
    #         category
    #   2. 'category': sums statistics for all activities into their sub
    #         category
    #   3. 'activity' (default): sums statistics for individual applications /
    #         web sites / activities
    #   4. 'productivity': productivity calculation (@see #productivity)
    #   5. 'efficiency': efficiency calculation (not applicable in "rank"
    #         perspective, @see #efficiency)
    # @option options [String] :by
    #   Lets you set the perspective of your report
    #   1. 'rank' (default): returns a ranked report of activities by total time spent
    #   2. 'time': returns a chronological report of activities
    #   3. 'member': returns an activity report grouped by member
    # @option options [String] :date
    #   Lets you set a single date for your report in a 'YYYY-MM-DD' format. Not
    #   valid with :to or :from. Cannot be a future date.
    # @option options [String] :from
    #   Lets you set a start date for your report in a 'YYYY-MM-DD' format. Valid
    #   with the :to option or alone (:to defaults to current date). Cannot be a
    #   future date or after the :to date (if supplied).
    # @option options [String] :to
    #   Lets you set an end date for your report in a 'YYYY-MM-DD' format. If a
    #   :to is supplied, :from must be supplied as well. Cannot be a future date or
    #   before the :from date.
    # @option options [String] :interval
    #   Lets you set the time interval for your report
    #   (ie. client.efficiency(interval:'day') returns the efficiency
    #   report by day.) Possible values include:
    #   1. 'minute': returns data in 5-minute increments
    #   2. 'hour' (default): returns data in 1-hour increments
    #   3. 'day': returns data in 1-day increments
    #   4. 'week': returns data in 1-week increments
    #   5. 'month': returns data in 1-month increments
    # @option options [String] :format
    #   Lets you specify a return type for your report
    #   1. default: returns an array of hashes with symbolized keys
    #   2. 'csv': returns a Mime::CSV object
    #
    # @return [Array<Hash>]
    #
    # @raise [Rescuetime::MissingCredentials] if the Rescuetime::Client has
    #   no set api key
    # @raise [Rescuetime::InvalidCredentials] if the provided api key is
    #   rejected by RescueTime
    # @since v0.1.0
    def activities(options = {})
      response = self.get BASE_URL, options

      case options[:format]
      when 'csv' then CSV.new(response.body, headers: true)
      else array_of_hashes_from_csv(response.body)
      end
    end

    # Returns efficiency report. Equivalent to
    #   #activities(by: time, detail: efficiency)
    # @see #activities valid options
    #
    # @param [Hash] options options hash (same as #activities)
    # @return [Array<Hash>]
    def efficiency(options = {})
      self.activities({by: 'time'}.merge(options.merge(detail: 'efficiency')))
    end

    # Returns productivity report. Equivalent to #activities(detail: productivity)
    # @see #activities valid options
    #
    # @example
    #   @client.productivity    # Equivalent to @client.activities(detail: 'productivity')
    #   # =>  [
    #   #       { :rank=>1, :time_spent_seconds=>6956, :number_of_people=>1, :productivity=>2 },
    #   #       { :rank=>2, :time_spent_seconds=>2635, :number_of_people=>1, :productivity=>-2 },
    #   #       { :rank=>3, :time_spent_seconds=>2415, :number_of_people=>1, :productivity=>1 },
    #   #       { :rank=>4, :time_spent_seconds=>1210, :number_of_people=>1, :productivity=>0 },
    #   #       { :rank=>5, :time_spent_seconds=>93, :number_of_people=>1, :productivity=>-1 }
    #   #     ]
    #
    # @param [Hash] options options hash (same as #activities)
    # @return [Array<Hash>]
    def productivity(options = {})
      self.activities(options.merge(detail: 'productivity'))
    end

    # Returns map of numeric productivity levels to meaning
    #
    # @example
    #   @activity = @client.activities(restrict_kind: 'productivity')[0]
    #   @primary_productivity = @client.productivity_levels[@activity[:productivity]]
    #
    #   puts "I have spent most of my time being #{@primary_productivity}."
    #   # => I have spent most of my time being Very Productive.
    #
    # @return [Hash] productivity levels, with integers from -2 to 2 as keys
    # @since v0.2.0
    def productivity_levels
      PRODUCTIVITY_LEVELS
    end

    private

    # Takes a CSV with headers and returns an array of hashes
    #
    # @param body[CSV] the original CSV file
    # @return [Array] an array of hashes, containing keys of the CSV headers
    # @since v0.1.0
    def array_of_hashes_from_csv(body)
      activities = CSV.new(body,
                           headers: true,
                           header_converters: :symbol,
                           converters: :all)

      activities.to_a.map(&:to_hash)
    end
  end
end
